src/java/org/apache/cassandra/security/AbstractCryptoProvider.java

```diff
@@ -0,0 +1,144 @@
+
+    /**
+     * Returns a runnable which installs this crypto provider.
+     *
+     * @return runnable which installs this provider
+     */
+    protected abstract Runnable installator() throws Exception;
+
+    /**
+     * Returns boolean telling if this provider was installed properly.
+     *
+     * @return {@code true} if provider was installed properly, {@code false} otherwise.
+     */
+    protected abstract boolean isHealthyInstallation() throws Exception;
+
+    /**
+     * The default installation runs {@link AbstractCryptoProvider#installator()} and after that
+     * {@link AbstractCryptoProvider#isHealthyInstallation()}.
+     * <p>
+     * If any step fails, it will not throw an exception unless the parameter
+     * {@link AbstractCryptoProvider#FAIL_ON_MISSING_PROVIDER_KEY} is {@code true}.
+     */
+    public void install() throws Exception
+    {
+        if (NoOpCryptoProvider.class.getName().equals(getProviderClassAsString()))
+        {
+            logger.debug("Installation of a crypto provider was skipped.");
+            return;
+        }
+
+        String failureMessage = null;
+        try
+        {
+            Class.forName(getProviderClassAsString());
+
+            Provider[] providers = Security.getProviders();
+
+            if (providers.length > 0 && providers[0] != null && providers[0].getName().equals(getProviderName()))
+            {
+                logger.debug("{} was already installed", getProviderName());
+                return;
+            }
+            else
+            {
+                installator().run();
+            }
+
+            if (isHealthyInstallation())
+            {
+                logger.info("{} successfully passed the healthiness check", getProviderName());
+            }
+            else
+            {
+                failureMessage = format("%s is not the highest priority provider. " +
+                                        "The most probable cause is that Cassandra node is not running on the same architecture " +
+                                        "the provider library is for. Please place the architecture-specific library " +
+                                        "for %s to the classpath and try again. ",
+                                        getProviderName(),
+                                        getProviderClassAsString());
+            }
+        }
+        catch (ClassNotFoundException ex)
+        {
+            failureMessage = getProviderClassAsString() + " is not on the class path!";
+        }
+        catch (Exception e)
+        {
+            failureMessage = format("The installation of %s was not successful, reason: %s",
+                                    getProviderClassAsString(), e.getMessage());
+        }
+
+        if (failureMessage != null)
+            if (failOnMissingProvider)
+                throw new ConfigurationException(failureMessage);
+            else
+                logger.warn(failureMessage);
```

Comment:
braces on the outer if ?

Reply:
@michaelsembwever I am getting different opinion on this from different people. You say this, somebody else says that I either braces everywhere or no braces at all.

Reply:
my understanding is no-braces is only when there is one line that idents at the next level (here there is two)

Reply:
No worries I ll change that. I think that will be better, I am uncomfortable to have not have them there even it works ...